pkg/types/types.go

```diff
@@ -197,29 +201,58 @@ type Metric struct {
 }
 
 // MergeMetrics merges metrics by name.
-func MergeMetrics(metrics [][]Metric) []Metric {
+// It returns merged metrics, number of rendered data points for the returned metrics,
+// and number of mismatched data points seen (if mismatchCheck is true).
+func MergeMetrics(metrics [][]Metric, mismatchCheck bool) ([]Metric, int, int) {
 	if len(metrics) == 0 {
-		return nil
+		return nil, 0, 0
 	}
 
 	if len(metrics) == 1 {
-		return metrics[0]
+		pointCount := 0
+		ms := metrics[0]
+		for _, m := range ms {
+			pointCount += len(m.Values)
+		}
+		return ms, pointCount, 0
 	}
 
-	names := make(map[string][]Metric)
+	metricByNames := make(map[string][]Metric)
 
 	for _, ms := range metrics {
 		for _, m := range ms {
-			names[m.Name] = append(names[m.Name], m)
+			metricByNames[m.Name] = append(metricByNames[m.Name], m)
 		}
 	}
 
 	merged := make([]Metric, 0)
-	for _, ms := range names {
-		merged = append(merged, mergeMetrics(ms))
+	pointCount := 0
+	mismatchCount := 0
+	var mismatchedMetricReport []map[string]interface{}
+	for _, ms := range metricByNames {
+		m, c := mergeMetrics(ms, mismatchCheck)
+		if c > 0 && len(mismatchedMetricReport) < mismatchedMetricReportLimit {
+			mismatchedMetricReport = append(mismatchedMetricReport, map[string]interface{}{
+				"metric_name":       m.Name,
+				"start":             m.StartTime,
+				"stop":              m.StopTime,
+				"step":              m.StepTime,
+				"mismatched_points": c,
+			})
+		}
+		merged = append(merged, m)
+		mismatchCount += c
+		pointCount += len(m.Values)
 	}
 
-	return merged
+	if mismatchCheck && mismatchCount > 0 {
+		corruptionLogger.Warn("metric mismatch observed",
+			zap.Any("mismatched_metrics", mismatchedMetricReport),
```

Comment:
`replicaMismatch` is kinda clear w/o context. Simply `mismatch` is not much better than `inconsistency` imho.